pkg/jobs/wait.go

```diff
@@ -0,0 +1,229 @@
+// created by this registry and thus are pre-claimed by it. This bypasses
+// the loop to discover jobs already claimed by this registry. If the jobs
+// turn out to not be claimed by this registry, it's not a problem.
+func (r *Registry) NotifyToResume(ctx context.Context, jobs ...jobspb.JobID) {
+	m := newJobIDSet(jobs...)
+	_ = r.stopper.RunAsyncTask(ctx, "resume-jobs", func(ctx context.Context) {
+		r.withSession(ctx, func(ctx context.Context, s sqlliveness.Session) {
+			r.filterAlreadyRunningAndCancelFromPreviousSessions(ctx, s, m)
+			r.resumeClaimedJobs(ctx, s, m)
+		})
+	})
+}
+
+// WaitForJobs waits for a given list of jobs to reach some sort
+// of terminal state.
+func (r *Registry) WaitForJobs(
+	ctx context.Context, ex sqlutil.InternalExecutor, jobs []jobspb.JobID,
+) error {
+	log.Infof(ctx, "waiting for %d %v queued jobs to complete", len(jobs), jobs)
+	done, cleanup := r.installWaitingSet(jobs...)
+	defer cleanup()
+	return r.waitForJobs(ctx, ex, jobs, done)
+}
+
+func (r *Registry) waitForJobs(
+	ctx context.Context, ex sqlutil.InternalExecutor, jobs []jobspb.JobID, done <-chan struct{},
+) error {
+
+	if len(jobs) == 0 {
+		return nil
+	}
+	buf := bytes.Buffer{}
+	for i, id := range jobs {
+		if i > 0 {
+			buf.WriteString(",")
+		}
+		buf.WriteString(fmt.Sprintf(" %d", id))
+	}
+	// Manually retry instead of using SHOW JOBS WHEN COMPLETE so we have greater
+	// control over retries. Also, avoiding SHOW JOBS prevents us from having to
+	// populate the crdb_internal.jobs vtable.
+	query := fmt.Sprintf(
+		`SELECT count(*) FROM system.jobs WHERE id IN (%s)
+       AND (status != $1 AND status != $2 AND status != $3 AND status != $4)`,
+		buf.String())
+	start := timeutil.Now()
+	ret := retry.StartWithCtx(ctx, retry.Options{
+		InitialBackoff: 500 * time.Millisecond,
+		MaxBackoff:     3 * time.Second,
+		Multiplier:     1.5,
+		Closer:         done,
+	})
+	ret.Next() // wait at least one InitialBackoff
+	for ret.Next() {
+		// We poll the number of queued jobs that aren't finished. As with SHOW JOBS
+		// WHEN COMPLETE, if one of the jobs is missing from the jobs table for
+		// whatever reason, we'll fail later when we try to load the job.
+		row, err := ex.QueryRowEx(
+			ctx,
+			"poll-show-jobs",
+			nil, /* txn */
+			sessiondata.InternalExecutorOverride{User: security.RootUserName()},
+			query,
+			StatusSucceeded,
+			StatusFailed,
+			StatusCanceled,
+			StatusRevertFailed,
+		)
+		if err != nil {
+			return errors.Wrap(err, "polling for queued jobs to complete")
+		}
+		if row == nil {
+			return errors.New("polling for queued jobs failed")
+		}
+		count := int64(tree.MustBeDInt(row[0]))
+		if log.V(3) {
+			log.Infof(ctx, "waiting for %d queued jobs to complete", count)
+		}
+		if count == 0 {
+			break
+		}
+	}
+	defer func() {
+		log.Infof(ctx, "waited for %d %v queued jobs to complete %v",
+			len(jobs), jobs, timeutil.Since(start))
+	}()
+	for i, id := range jobs {
+		j, err := r.LoadJob(ctx, id)
```

Comment:
i wonder: at what point would just starting a rangefeed and filtering for things we care about be faster?
Certainly would bypass any locking issues, and presumably if the job set cardinality is sufficiently large, running the above count(*) query plus a load for each job might be pretty expensive.